src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,189 @@
+                log_warning("'%s' changed sysctl '%s' from '%s' to '%s', we want '%s'", we->comm, we->name, we->current, we->newvalue, value);
+
+        return 0;
+}
+
+static int on_ringbuf_io(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        struct ring_buffer *rb = ASSERT_PTR(userdata);
+        int r;
+
+        r = sym_ring_buffer__poll(rb, 1000);
+        if (r < 0 && errno != EINTR)
+                log_error_errno(errno, "Error polling ring buffer: %m");
+
+        return 0;
+}
+
+int sysctl_add_monitor(Manager *manager) {
+        _cleanup_(sysctl_monitor_bpf_freep) struct sysctl_monitor_bpf *obj = NULL;
+        _cleanup_(bpf_link_freep) struct bpf_link *sysctl_link = NULL;
+        _cleanup_(rb_freep) struct ring_buffer *sysctl_buffer = NULL;
+        _cleanup_close_ int cgroup_fd = -EBADF, rootcg = -EBADF;
+        int idx = 0;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0)
+                return log_info_errno(r, "sysctl monitor disabled, as BPF support is not available.");
+
+        rootcg = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/");
+        if (rootcg < 0)
+                return log_warning_errno(rootcg, "Failed to open cgroup: %m");
+
+        obj = sysctl_monitor_bpf__open_and_load();
+        if (!obj)
+                return log_warning_errno(errno, "Failed to load sysctl monitor BPF program: %m");
+
+        cgroup_fd = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/system.slice/systemd-networkd.service");
+        if (cgroup_fd < 0)
+                return log_warning_errno(cgroup_fd, "Failed to open cgroup: %m");
+
+        if (sym_bpf_map_update_elem(sym_bpf_map__fd(obj->maps.cgroup_map), &idx, &cgroup_fd, BPF_ANY))
+                return log_error_errno(errno, "Failed to update cgroup map: %m");
+
+        sysctl_link = sym_bpf_program__attach_cgroup(obj->progs.sysctl_monitor, rootcg);
+        r = bpf_get_error_translated(sysctl_link);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to attach sysctl monitor BPF program to cgroup: %m");
+
+        sysctl_buffer = sym_ring_buffer__new(
+                        sym_bpf_map__fd(obj->maps.written_sysctls),
+                        sysct_write_event, &manager->sysctl_shadow, NULL);
+        if (!sysctl_buffer)
+                return log_warning_errno(errno, "Failed to create ring buffer: %m");
+
+        r = sd_event_add_io(manager->event, &manager->sysctl_event_source,
+                        sym_ring_buffer__epoll_fd(sysctl_buffer), EPOLLIN, on_ringbuf_io, sysctl_buffer);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to watch sysctl event ringbuffer: %m");
+
+        manager->sysctl_link = TAKE_PTR(sysctl_link);
+        manager->sysctl_skel = TAKE_PTR(obj);
+        manager->sysctl_buffer = TAKE_PTR(sysctl_buffer);
+        manager->cgroup_fd = TAKE_FD(cgroup_fd);
+
+        return 0;
+}
+
+int sysctl_remove_monitor(Manager *manager) {
```

Comment:
functions that destroy things should typically not have the permission to fail, i.e. they should return void.

they should close/destroy what is opened, be idempotent, and never fail